src/ssa.c

```diff
@@ -0,0 +1,1140 @@
+void forward_bb_traversal(fn_t *fn,
+                          basic_block_t *bb,
+                          void (*preorder_callback)(fn_t *, basic_block_t *),
+                          void (*postorder_callback)(fn_t *, basic_block_t *))
+{
+    bb->visited++;
+
+    if (preorder_callback)
+        preorder_callback(fn, bb);
+
+    if (bb->next && bb->next->visited < fn->visited)
+        forward_bb_traversal(fn, bb->next, preorder_callback,
+                             postorder_callback);
+    if (bb->then_ && bb->then_->visited < fn->visited)
+        forward_bb_traversal(fn, bb->then_, preorder_callback,
+                             postorder_callback);
+    if (bb->else_ && bb->else_->visited < fn->visited)
+        forward_bb_traversal(fn, bb->else_, preorder_callback,
+                             postorder_callback);
+
+    if (postorder_callback)
+        postorder_callback(fn, bb);
+}
+
+void backward_bb_traversal(fn_t *fn,
```

Comment:
To avoid potential naming conflicts, I would like to propose consistent naming scheme.
* basic block facilities: starting with `bb_`. e.g., `bb_forward_traversal`
* dominator tree: starting with `dom_`. e.g., `dom_build`.